feat(developer): implement import and implied 🙀

srl295 opened the pull request:
- k_100, k_101, k_102 all failed for different reasons. 
- this PR adds support for import to common/web/types
- this also supports implied
- two of the three keyboards have had the following added, so that they now pass.

```xml
    <import base="cldr" path="techpreview/keys-Latn-implied.xml" />
```

For: 
- #7401
- #7534
- #7964

@keymanapp-test-bot skip

mcdurdin commented on core/tests/unit/ldml/keyboards/k_100_fail.xml:
```diff
@@ -0,0 +1,43 @@
+</keyboard>
+
+<!--
+  file:///Users/srl295/src/keymanapp/keyman/common/web/types/build/src/kvk/kvk-file-writer.js:62
+        str.len = value.length + 1;
```
* `gap` not yet supported (would be coming with touch layouts)

srl295 replied:
It's actually "key without to="

srl295 replied:
now supported

srl295 commented:
just updating the brokenness !

srl295 commented:
k_100 now passes, thanks to #7955

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -1,9 +1,16 @@
-import { LDMLKeyboardXMLSourceFile } from './ldml-keyboard-xml.js';
+import { LDMLKeyboardXMLSourceFile, LKImport } from './ldml-keyboard-xml.js';
 import Ajv from 'ajv';
+import * as fs from 'fs';
 import { boxXmlArray } from '../util/util.js';
 
 export default class LDMLKeyboardXMLSourceFileReader {
+  readImportFile(version: string, subpath: string): Buffer {
+    // TODO-LDML: sanitize input string
+    let importPath = new URL(`../import/${version}/${subpath}`, import.meta.url);
+    return fs.readFileSync(importPath);
+  }
```
File operations should be done through [`CompilerCallbacks`](https://github.com/keymanapp/keyman/blob/cb754353175f515e3677670ae1bf8776b7ce8f1c/developer/src/kmc/src/kmc.ts#L41-L57). This will allow us to do web-side compiler, as well as simplifying mocking for tests.

srl295 replied:
done.

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -39,26 +46,76 @@ export default class LDMLKeyboardXMLSourceFileReader {
           boxXmlArray(obj, key);
+          // Now, resolve the import
+          this.resolveImports(obj, subtag);
+          // now delete the import array we so carefully constructed, the caller does not
+          // want to see it.
```
🤣

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -39,26 +46,76 @@ export default class LDMLKeyboardXMLSourceFileReader {
+  }
+
+  private resolveImports(obj: any, subtag: string) {
+    for (const anImport /* : LKImport */ of obj['import']) {
+      const asImport: LKImport = anImport;
```
It's not possible to cast the `obj['import']` to avoid the duplicate variables?

Something like:

```
for (const anImport of (obj['import'] as LKImport[]) {
```

srl295 replied:
that works, thanks

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -39,26 +46,76 @@ export default class LDMLKeyboardXMLSourceFileReader {
+        throw new Error(`import element with base ${asImport.base} is unsupported.`);
+      }
+      const paths = asImport.path.split('/');
+      if (!paths[0] || !paths[1] || paths.length !== 2) {
```
Testing zero-length strings? Might be clearer to do:

```suggestion
      if (paths[0] == '' || paths[1] == '' || paths.length !== 2) {
```

srl295 replied:
thanks

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -39,26 +46,76 @@ export default class LDMLKeyboardXMLSourceFileReader {
+    for (const anImport /* : LKImport */ of obj['import']) {
+      const asImport: LKImport = anImport;
+      if (asImport.base !== 'cldr') {
+        throw new Error(`import element with base ${asImport.base} is unsupported.`);
```
Let's use a subclass of `CompilerError` called `CompilerImportResolveError` so that we can handle it outside the loader later

srl295 replied:
taken as a TODO-LDML

srl295 replied:
@mcdurdin question.  For the followon PR… should it *throw* or should it *report a message* if it can?

srl295 replied:
now that the callback for message reporting is available to this code

mcdurdin replied:
Report a message if possible -- a compiler should throw only unrecoverable or internal errors, because we'd like to continue and learn about more errors later in the file. (Of course, some errors (e.g. missing `}` in C) are going to result in a cascade of disaster but that's normal.)

mcdurdin commented on core/tests/unit/ldml/keyboards/k_100_keytest.xml:
```diff
@@ -0,0 +1,48 @@
+  </layers>
+</keyboard>
+
+<!--
+  file:///Users/srl295/src/keymanapp/keyman/common/web/types/build/src/kvk/kvk-file-writer.js:62
+        str.len = value.length + 1;
+                        ^
+
+TypeError: Cannot read properties of undefined (reading 'length')
+    at KvkFileWriter.setString (file:///Users/srl295/src/keymanapp/keyman/common/web/types/build/src/kvk/kvk-file-writer.js:62:25)
+    at KvkFileWriter.build (file:///Users/srl295/src/keymanapp/keyman/common/web/types/build/src/kvk/kvk-file-writer.js:48:18)
+    at KvkFileWriter.write (file:///Users/srl295/src/keymanapp/keyman/common/web/types/build/src/kvk/kvk-file-writer.js:9:29)
+    at compileKeyboard (file:///Users/srl295/src/keymanapp/keyman/developer/src/kmc/build/src/kmc.js:73:31)
+    at file:///Users/srl295/src/keymanapp/keyman/developer/src/kmc/build/src/kmc.js:82:18
+    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
+    at async Promise.all (index 0)
+    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
+    at async loadESM (node:internal/process/esm_loader:91:5)
+    at async handleMainPromise (node:internal/modules/run_main:65:12)
+-->
```
```suggestion
```

❓❓❓

srl295 replied:
junk, will delete.

srl295 replied:
this conversation is officially unresolvable :)

mcdurdin replied:
💯

mcdurdin commented on core/tests/unit/ldml/keyboards/k_102_keytest.xml:
```diff
@@ -0,0 +1,43 @@
+
+</keyboard>
+
+<!--
+'keys' is actually optional.
+
+
+ERROR 300007: The source file has an invalid structure: Error: data/keyboard must have required property 'keys'
+file:///Users/srl295/src/keymanapp/keyman/developer/src/kmc/build/src/kmc.js:82
+let [kmx, kvk] = compileKeyboard(inputFilename, options);
+                 ^
+
+TypeError: compileKeyboard is not a function or its return value is not iterable
+    at file:///Users/srl295/src/keymanapp/keyman/developer/src/kmc/build/src/kmc.js:82:18
+    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
+    at async Promise.all (index 0)
+    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
+    at async loadESM (node:internal/process/esm_loader:91:5)
+    at async handleMainPromise (node:internal/modules/run_main:65:12)
+  -->
```
```suggestion
```

❔❔❔

mcdurdin commented on developer/src/kmc/build.sh:
```diff
@@ -46,6 +46,16 @@ else
   # We need the schema file at runtime and bundled, so always copy it for all actions except `clean`
   mkdir -p "$THIS_SCRIPT_PATH/build/src/"
   cp "$KEYMAN_ROOT/resources/standards-data/ldml-keyboards/techpreview/ldml-keyboard.schema.json" "$THIS_SCRIPT_PATH/build/src/"
+  # Store imports
+  # load all versions that have a cldr_info.json
+  for CLDR_INFO_PATH in "$KEYMAN_ROOT/resources/standards-data/ldml-keyboards/"*/cldr_info.json
+  do
+    CLDR_PATH=$(dirname "$CLDR_INFO_PATH")
+    CLDR_VER=$(basename "$CLDR_PATH")
+    mkdir -p "$THIS_SCRIPT_PATH/build/src/import/$CLDR_VER"
+    # TODO-LDML: When these are copied, the DOCTYPE will break due to the wrong path.
+    cp "$CLDR_INFO_PATH" "$CLDR_PATH/import/"*.xml "$THIS_SCRIPT_PATH/build/src/import/$CLDR_VER/"
+  done
```
This is duplicated from common/web/types/build.sh. I'm not sure why we need both?

srl295 replied:
it should be one or the other.  depending on where imports land.

srl295 replied:
updated to remove from kmc

mcdurdin reviewed (changes requested):
This is looking pretty good. However, the actual file load should be implemented as a `CompilerCallback`. That means we should be passing the compiler callback class into the xml loader, which may mean we can use it also for error reporting, which would be nicer than throwing errors (especially if they are recoverable errors so we can continue to log additional errors later in the file, like most compilers do)

srl295 commented:
OK. Whew. We still pass, even with lots of merging.  Now back to refactoring work!

srl295 commented:
> CompilerCallback

I'm thinking now that the import / implied should also be done in a CompilerCallback

mcdurdin commented:
Oh, please don't merge approved PRs into unapproved PRs -- it makes the review process quite horrible 😢

srl295 commented:
> Oh, please don't merge approved PRs into unapproved PRs -- it makes the review process quite horrible 😢

? I don't think I did that… 

mcdurdin commented on common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:
```diff
@@ -99,13 +107,12 @@ export default class LDMLKeyboardXMLSourceFileReader {
     // now, the explicit imports
-    for (const anImport /* : LKImport */ of obj['import']) {
-      const asImport: LKImport = anImport;
+    for (const asImport of (obj['import'] as LKImport[])) {
```
:+1:

srl295 replied:
thx for the tip

mcdurdin reviewed (approved):
LGTM I think

mcdurdin commented:
> I hear you, but ? I don't think I did that… 
> 
> I merged an unreviewed PR here ( #8107 ) which had a refactoring that addressed some of the comments.
> 
> In between was a PR #8106 which had some API refactoring. That one is now the parent of this one, since it comes out separately.
> 
> The other merged PRs were approved and merged directly into feature-ldml

Ah, okay, I see. It was the merge of #8107 which confused me -- as we don't require reviews of PRs that are not to locked branches, this is certainly possible :grin:

srl295 commented:
> > I hear you, but ? I don't think I did that… 
> > I merged an unreviewed PR here ( #8107 ) which had a refactoring that addressed some of the comments.
> > In between was a PR #8106 which had some API refactoring. That one is now the parent of this one, since it comes out separately.
> > The other merged PRs were approved and merged directly into feature-ldml
> 
> Ah, okay, I see. It was the merge of #8107 which confused me -- as we don't require reviews of PRs that are not to locked branches, this is certainly possible 😁

I'm trying (and not doing so well i suppose!) to keep PRs small and reviewable.  A couple of times I've realized that having a separate PR is counterproductive  - it would ask us to review and merge the 'bad' parent PR and then later reviewe and merge the 'good' child PR.  This is a case in point.

mcdurdin commented:
> I'm trying (and not doing so well i suppose!) to keep PRs small and reviewable. A couple of times I've realized that having a separate PR is counterproductive - it would ask us to review and merge the 'bad' parent PR and then later reviewe and merge the 'good' child PR. This is a case in point.

'Tis all good. The occasional bit of confusion is part of the process.

> _edit_ so the ask is, is there a way I can better annotate #8107 - something like "merged by author for shuffling purposes*?

I think a note (the same note) in both PRs would be helpful in this case. How about something like this, capturing (hopefully) reason behind the merge, and highlighting the decision made:

* I went ahead and merged #8107 into #7402 because #7402 was broken without the changes in #8107, and turns out it didn't really make sense to try and review the two PRs separately.

mcdurdin commented:
![image](https://user-images.githubusercontent.com/4498365/215234714-ba90008d-bd27-48ae-8cc5-ea770aae0b73.png)

And sometimes GH's annotations aren't helpful -- when a PR is referenced in a dot point, it turns `#xxxx` into a title link...

srl295 commented:
:+1: on asking to clear up confusion, thanks!

> I think a note (the same note) in both PRs would be helpful in this case.

sounds good…

> And sometimes GH's annotations aren't helpful

Yes, it just looks like a bunch of screaming cats visiting  a link farm 🤣
